builder/src/prepare_project.py

```diff
@@ -151,7 +291,263 @@ def supports(self, version):
+                  'supported app.')
+            sys.exit(1)
+
+        base_image = get_base_image(version_map, minor_version)
+        if base_image is None:
+            print('The app requires .NET Core runtime version {0} which is not supported at ' +
+                  'this time.').format(minor_version)
+            sys.exit(1)
+
+        project_name = self._get_project_assembly_name()
+        contents = SingleProjectApp.DOCKERFILE_CONTENTS.format(runtime_image=base_image.image,
+                                                               dll_name=project_name)
+        with open(output, 'wt') as out:
+            out.write(contents)
+
+    def _get_project_assembly_name(self):
+        """Returns the name of the main assembly for the project.
+
+        This method will use the file path as the basis for the name
+        of the assembly.
+
+        Returns:
+            A string with the name of the main assbemly for the project.
+        """
+        basename = os.path.basename(self.project)
+        return os.path.splitext(basename)[0]
+
+    def _get_project_runtime_version(self):
+        """This method returns runtime targeted by this project.
+
+        This method will parse the .csproj file and look for the
+        TargetFramework property if present.
+
+        Returns:
+            A string with the target version.
+        """
+        tree = ET.parse(self.project)
+        root = tree.getroot()
+        framework_element = root.find('./PropertyGroup/TargetFramework')
+        if framework_element is None:
+            return None
+
+        framework = framework_element.text
+        if not framework.startswith(NETCOREAPP_VERSION_PREFIX):
+            print('The app is not supported to release, must be an executable.')
+            sys.exit(1)
+
+        return framework[len(NETCOREAPP_VERSION_PREFIX):]
+
+
+class SolutionApp(SingleProjectApp):
+    """An app that is composed of a solution and one, or more, projects.
+
+    This class contains the information for an app that is composed of
+    a .sln and one, or more, projects. One of the projects will be the
+    startup project and thus the one used to launch the app in the
+    generated Dockerfile.
+    """
+
+    # Dockerfile template to be used when packaging .sln based apps.
+    DOCKERFILE_CONTENTS = textwrap.dedent(
+        """\
+        FROM gcr.io/cloud-builders/csharp/dotnet AS builder
+        COPY . /src
+        WORKDIR /src
+        RUN dotnet restore --packages /packages
+        RUN dotnet publish -c Release -o /published {main_project}
+
+        FROM {runtime_image}
+        COPY --from=builder /published /app
+        ENV ASPNETCORE_URLS=http://*:${{PORT}}
+        WORKDIR /app
+        ENTRYPOINT [ "dotnet", "{dll_name}.dll" ]
+        """)
+
+    def __init__(self, root, app_yaml):
+        main_project = get_startup_project(app_yaml)
+        super(SolutionApp, self).__init__(root, get_startup_project(app_yaml))
```

Comment:
Why are we calling `get_startup_project` a second time?

Reply:
Done.